### ISSUE TITLE
Greasemonkey scripts which rely on shared global state don't work in qutebrowser

### ISSUE THREAD

**Issue by LiteracyFanatic** [qutebrowser/qutebrowser#5266]
**Version info**:
qutebrowser v1.10.1
Git commit: 
Backend: QtWebEngine (Chromium 77.0.3865.129)
Qt: 5.14.1

**Description**
I am trying to use the [Wani Kani Double-Check](https://greasyfork.org/en/scripts/35063-wanikani-double-check) script for [WaniKani](https://www.wanikani.com), a website for studying Japanese Kanji. This script and many others written for the same site rely on the  [Wani Kani Open Framework](https://greasyfork.org/en/scripts/38582-wanikani-open-framework) script. The scripts fail with the following error messages:

```
Uncaught TypeError: Cannot set property 'Settings' of undefined
    at userscript:GM-Wanikani Open Framework - Settings module:231
    at userscript:GM-Wanikani Open Framework - Settings module:806
    at userscript:GM-Wanikani Open Framework - Settings module:811

Uncaught TypeError: Cannot set property 'Progress' of undefined
    at userscript:GM-Wanikani Open Framework - Progress module:210
    at userscript:GM-Wanikani Open Framework - Progress module:392
    at userscript:GM-Wanikani Open Framework - Progress module:397
```

I'll try to summarize, to the best of my knowledge, what is supposed to happen and why it isn't working. The Open Framework script creates a property on the global `window` called `wkof` which it uses to expose its functionality to other scripts. It relies on several different modules which it fetches when it is run. Each of these modify the global `wkof` object. Finally, Double-Check and other scripts access the Open Framework API through `window.wkof` and use it to implement their functionality.

```
	var supported_modules = {
		Apiv2:    { url: 'https://greasyfork.org/scripts/38581-wanikani-open-framework-apiv2-module/code/Wanikani%20Open%20Framework%20-%20Apiv2%20module.js?version=747866'},
		ItemData: { url: 'https://greasyfork.org/scripts/38580-wanikani-open-framework-itemdata-module/code/Wanikani%20Open%20Framework%20-%20ItemData%20module.js?version=767868'},
		Menu:     { url: 'https://greasyfork.org/scripts/38578-wanikani-open-framework-menu-module/code/Wanikani%20Open%20Framework%20-%20Menu%20module.js?version=745388'},
		Progress: { url: 'https://greasyfork.org/scripts/38577-wanikani-open-framework-progress-module/code/Wanikani%20Open%20Framework%20-%20Progress%20module.js?version=601473'},
		Settings: { url: 'https://greasyfork.org/scripts/38576-wanikani-open-framework-settings-module/code/Wanikani%20Open%20Framework%20-%20Settings%20module.js?version=607871'},
	};
```

```
	var published_interface = {
		include: include,              // include(module_list)        => Promise
		ready:   ready,                // ready(module_list)          => Promise

		load_file:   load_file,        // load_file(url, use_cache)   => Promise
		load_css:    load_css,         // load_css(url, use_cache)    => Promise
		load_script: load_script,      // load_script(url, use_cache) => Promise

		file_cache: {
			dir:    {},                // Object containing directory of files.
			ls:     file_cache_list,   // ls()
			clear:  file_cache_clear,  // clear()             => Promise
			delete: file_cache_delete, // delete(name)        => Promise
			flush:  file_cache_flush,  // flush()             => Promise
			load:   file_cache_load,   // load(name)          => Promise
			save:   file_cache_save,   // save(name, content) => Promise
			no_cache:file_nocache,     // no_cache(modules)
		},

		on:      wait_event,           // on(event, callback)
		trigger: trigger_event,        // trigger(event[, data1[, data2[, ...]]])

		get_state:  get_state,         // get(state_var)
		set_state:  set_state,         // set(state_var, value)
		wait_state: wait_state,        // wait(state_var, value[, callback[, persistent]]) => if no callback, return one-shot Promise

		version: {
			value: version,
			compare_to: compare_to,    // compare_version(version)
		}
	};
```

```
	function startup() {
		global.wkof = published_interface;

		// Mark document state as 'ready'.
		if (document.readyState === 'complete') {
			doc_ready();
		} else {
			window.addEventListener("load", doc_ready, false);  // Notify listeners that we are ready.
		}

		// Open cache, so wkof.file_cache.dir is available to console immediately.
		file_cache_open();
		wkof.set_state('wkof.wkof', 'ready');
	}
	startup();
```

This approach to sharing common functionality fails because qutebrowser wraps each script in a chunk of code which proxies the `window` object. Thus `window.wkof` is undefined as far as the other scripts are concerned.

Not sure what the best way to fix this would be. Off the top of my head I can think of a couple of possible approaches:

1. Provide a setting to allow scripts to write to the real `window` object instead of a proxy. This could apply globally or just to a list of whitelisted scripts.

2. Provide a mechanism for associating a group of scripts with the same `window` proxy. Perhaps a collection of named proxies could be kept in the global scope. Regular scripts would get assigned a random identifier or one based on their name. Groups of scripts that need a shared context could either be defined in a setting in qutebrowser's configuration file or through an annotation in the scripts themselves. Then the wrapper that qutebrowser injects the scripts into would check if the neccessary proxy already existed and pass it instead of creating a new one.

**How to reproduce**
Install the mentioned scripts, visit www.wanikani.com, and open the web console to view the error messages. `window.wkof` will be undefined in the console. In Firefox and other browsers it is defined in the global scope.

**Comment by The-Compiler** [comment]
Hmm, I'm surprised those scripts work in other Greasemonkey implementations. I thought they also isolated `window` in the way we do.

@toofar what are your thoughts?

**Comment by LiteracyFanatic** [comment]
I have no idea if the original Greasemonkey extension isolated window or not. I'm running these scripts under the [Tampermonkey](https://github.com/Tampermonkey/tampermonkey) extension which claims compatibility with Greasemonkey. The README seems to indicate that scripts have access to `window`. Looking through the sourcecode I found some unimplimented [TODO comments about private window contexts](https://github.com/Tampermonkey/tampermonkey/blob/master/src/environment.js).

**Comment by The-Compiler** [comment]
Greasemonkey has [unsafeWindow](https://wiki.greasespot.net/UnsafeWindow) (and so does qutebrowser) to access the real window object.

I guess the question is whether the `window` proxy object is shared between different userscripts with Greasemonkey or not. If it is, we should probably do the same. If not, this seems like a bug in the script and it only works in Tampermonkey coincidentally.

**Comment by LiteracyFanatic** [comment]
It looks like Greasemonkey uses a single `window` proxy. I created two small test scripts to check this.

```
// ==UserScript==
// @name     a
// ==/UserScript==

window.a = 'test';
```

```
// ==UserScript==
// @name     b
// ==/UserScript==

console.log(window.a);
```

This outputs 'test' to the console as expected.

It's also worth mentioning that execution order matters for related scripts. Tampermonkey provides a UI for this. 3.x versions of Greasemonkey also provided an explicit UI mechanism to set the execution order of scripts. 4.x simply executes scripts in alphabetical order. I'm personally fine with prefixing the script names in my file system to get the correct ordering. It looks like qutebrowser is currently loading the scripts with the `glob` module which can [produce different results depending on the user's file system](https://docs.python.org/3.8/library/glob.html), so it may be worth manually sorting the returned list before iterating over it for the sake of correctness.

All that being said, it appears that these changes alone will not be enough to make the WaniKani Open Framework script work since it won't run under Greasemonkey. It looks like it injects the extra modules it loads into script tags on the page. Those scripts get executed in the global scope and thus can't access `wkof`. This is really a problem with the design of the script itself since even Tampermonkey where it currently runs has indicated that it may proxy `window` in the future. I'll take that up with the author of the script. The single shared proxy is still needed for scripts depending on the Open Framework to function correctly.

In summary, I think qutebrowser should use a single proxied `window` object and load scripts in a deterministic order to more closely match the behaviour of other implementations.

**Comment by toofar** [comment]
Thanks for looking into it so thoroughly, I suppose we could put the proxy thing on window._qute_gm_window_proxy or something and re-use it since it isn't for security, just to stop accidentally overwriting things. Which this would open the door for more but oh well.

**Comment by LiteracyFanatic** [comment]
I just opened a pull request addressing the `window` proxy issue. It looks like the execution order will be more complicated to deal with though. See this Qt [issue](https://codereview.qt-project.org/c/qt/qtwebengine/+/281433). I suppose it would be possible to control the order with a wrapper script until support lands in Qt itself, but that feels like a bit of a hack and would make debugging in the web inspector more awkward. Thoughts?

Also, this doesn't really affect qutebrowser, but for the sake of correctness I should amend my previous claim that Tampermonkey doesn't use a sandbox. It turns out that it doesn't use a proxy when `@grant none` is set. This is the default if `@grant` isn't specified at all. In the case where Greasemonkey API functions are granted, Tampermonkey does sandbox its scripts.

**Comment by The-Compiler** [comment]
@LiteracyFanatic Do you think this can be closed now that #5271 is in? I don't really have much to say about the script order - I don't think it makes sense for qutebrowser to work around that.

**Comment by LiteracyFanatic** [comment]
Sounds good.
